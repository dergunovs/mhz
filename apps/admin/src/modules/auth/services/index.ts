import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
