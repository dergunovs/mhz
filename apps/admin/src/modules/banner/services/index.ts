import { ComputedRef, Ref } from 'vue';

import { api, useQuery, useMutation, IPageQuery, convertParams } from 'mhz-helpers';
import { API_BANNER, IBanner, IBaseReply, IFilterData } from 'mhz-contracts';

export function getBanners(query: Ref<IPageQuery | number>) {
  async function fn() {
    const params = convertParams(query);

    const { data } = await api.get<{ data: IBanner[]; total: number; filters: IFilterData }>(API_BANNER, { params });

    return data;
  }

  return useQuery({ queryKey: [API_BANNER, query], queryFn: fn });
}

export function getBanner(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: IBanner }>(`${API_BANNER}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_BANNER, id], queryFn: fn });
}

export function postBanner(options: object) {
  async function fn(formData: IBanner) {
    const { data } = await api.post<IBaseReply>(API_BANNER, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}

export function updateBanner(options: object) {
  async function fn(formData: IBanner) {
    const { data } = await api.patch<IBaseReply>(`${API_BANNER}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}

export function deleteBanner(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_BANNER}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}
