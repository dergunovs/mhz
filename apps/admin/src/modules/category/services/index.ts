import { ComputedRef } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import { API_CATEGORY, ICategory, IBaseReply } from 'mhz-contracts';

export function getCategories(options?: object) {
  async function fn() {
    const { data } = await api.get<{ data: ICategory[] }>(API_CATEGORY);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY], queryFn: fn, ...options });
}

export function getCategory(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: ICategory }>(`${API_CATEGORY}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY, id], queryFn: fn });
}

export function postCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.post<IBaseReply>(API_CATEGORY, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function updateCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.patch<IBaseReply>(`${API_CATEGORY}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function deleteCategory(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_CATEGORY}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}
