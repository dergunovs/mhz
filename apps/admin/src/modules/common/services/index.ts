import { Ref } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import {
  API_UPLOAD,
  API_UPLOAD_SINGLE,
  API_SEARCH,
  API_STATS_COUNT,
  ISearchResults,
  IEntitiesCount,
  IBaseReply,
} from 'mhz-contracts';

export function search(query: Ref<string>, isAdmin?: boolean) {
  async function fn() {
    const { data } = await api.get<ISearchResults>(API_SEARCH, { params: { search: query.value, isAdmin } });

    return data;
  }

  return useQuery({ queryKey: [API_SEARCH, query], queryFn: fn, enabled: false });
}

export function getEntitiesCount() {
  async function fn() {
    const { data } = await api.get<IEntitiesCount>(API_STATS_COUNT);

    return data;
  }

  return useQuery({ queryKey: [API_STATS_COUNT], queryFn: fn });
}

export function uploadFile(options: object, width?: string) {
  async function fn(file?: File) {
    if (!file) throw new Error();

    const formData = new FormData();

    formData.append('file', file);

    const { data } = await api.post<string>(API_UPLOAD_SINGLE, formData, { params: { width } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function uploadFiles(options: object, width?: string, thumb?: boolean) {
  async function fn(files: File[]) {
    const formData = new FormData();

    files.forEach((file) => formData.append('files', file));

    const { data } = await api.post<string[]>(API_UPLOAD, formData, { params: { width, thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function deleteFile(thumb?: boolean) {
  async function fn(filename: string) {
    const { data } = await api.delete<IBaseReply>(`${API_UPLOAD}/${filename}`, { params: { thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn });
}
