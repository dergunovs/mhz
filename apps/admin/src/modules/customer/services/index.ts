import { Ref, ComputedRef } from 'vue';

import { api, useQuery, IPageQuery, convertParams } from 'mhz-helpers';
import { API_CUSTOMER, ICustomer } from 'mhz-contracts';

export function getCustomers(query: Ref<IPageQuery | number>) {
  async function fn() {
    const params = convertParams(query);

    const { data } = await api.get<{ data: ICustomer[]; total: number }>(API_CUSTOMER, { params });

    return data;
  }

  return useQuery({ queryKey: [API_CUSTOMER, query], queryFn: fn });
}

export function getCustomer(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: ICustomer }>(`${API_CUSTOMER}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_CUSTOMER, id], queryFn: fn });
}
