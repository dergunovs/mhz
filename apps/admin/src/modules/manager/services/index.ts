import { ComputedRef, Ref } from 'vue';

import { IManager } from 'mhz-types';
import { api, useMutation, useQuery, IPageQuery } from 'mhz-helpers';

import { API_MANAGER } from '@/manager/constants';

export function getManagers(query: Ref<IPageQuery | number>) {
  async function fn(): Promise<{ data: IManager[]; total: number }> {
    const params =
      typeof query.value === 'number'
        ? { page: query.value }
        : {
            page: query.value.page || 1,
            sort: query.value.sort.value,
            dir: query.value.sort.isAsc === false ? 'desc' : 'asc',
          };

    const { data } = await api.get(API_MANAGER, { params });

    return data;
  }

  return useQuery({ queryKey: [API_MANAGER, query], queryFn: fn });
}

export function getManager(id?: ComputedRef<string | string[]>) {
  async function fn(): Promise<IManager | null> {
    if (!id?.value) return null;

    const { data } = await api.get(`${API_MANAGER}/${id.value}`);

    return data;
  }

  return useQuery({ queryKey: [API_MANAGER, id], queryFn: fn, refetchOnMount: true });
}

export function postManager(options: object) {
  async function fn(formData: IManager) {
    await api.post(API_MANAGER, formData);
  }

  return useMutation({ mutationKey: [API_MANAGER], mutationFn: fn, ...options });
}

export function updateManager(id: ComputedRef<string | undefined>, options: object) {
  async function fn(formData: IManager) {
    await api.patch(`${API_MANAGER}/${id.value}`, formData);
  }

  return useMutation({ mutationKey: [API_MANAGER, id], mutationFn: fn, ...options });
}

export function deleteManager(options: object) {
  async function fn(id?: string) {
    await api.delete(`${API_MANAGER}/${id}`);
  }

  return useMutation({ mutationKey: [API_MANAGER], mutationFn: fn, ...options });
}
