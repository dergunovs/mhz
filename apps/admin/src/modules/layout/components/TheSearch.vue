<template>
  <div>
    <UiSearch
      v-model="searchQuery"
      :searchScheme="SEARCH_SCHEME"
      :results="results"
      :isSuccess="isSuccess"
      lang="en"
      data-test="search"
    />
  </div>
</template>

<script setup lang="ts">
import { ref, watch } from 'vue';

import { UiSearch } from 'mhz-ui';

import { SEARCH_SCHEME } from '@/layout/constants';
import { search } from '@/common/services';

const searchQuery = ref('');

const { data: results, refetch, isSuccess } = search(searchQuery);

watch(
  () => searchQuery.value,
  () => {
    if (searchQuery.value.length > 2) refetch();
  }
);
</script>
