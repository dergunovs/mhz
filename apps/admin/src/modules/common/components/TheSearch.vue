<template>
  <div :class="$style.container">
    <UiInput v-model="searchQuery" :appendIcon="IconSearch" placeholder="Search" />

    <div v-if="searchQuery.length" :class="$style.results">
      <template v-if="searchQuery.length < 3 && !isResults">Please enter 3 or more symbols</template>
      <template v-if="searchQuery.length > 2 && !isSuccess">Loading...</template>
      <template v-if="searchQuery.length > 2 && !isResults && isSuccess">No results</template>

      <template v-if="searchQuery.length > 2 && isResults">
        <div v-if="results?.products.length" :class="$style.resultsInner">
          <div>Products:</div>
          <RouterLink
            v-for="product in results.products"
            :key="product._id"
            :to="`${URL_PRODUCT_EDIT}/${product._id}`"
            @click="clearSearch"
          >
            {{ product.title }}
          </RouterLink>
        </div>

        <div v-if="results?.categories.length" :class="$style.resultsInner">
          <div>Categories:</div>
          <RouterLink
            v-for="category in results.categories"
            :key="category._id"
            :to="`${URL_CATEGORY_EDIT}/${category._id}`"
            @click="clearSearch"
          >
            {{ category.title }}
          </RouterLink>
        </div>

        <div v-if="results?.manufacturers.length" :class="$style.resultsInner">
          <div>Manufacturers:</div>
          <RouterLink
            v-for="manufacturer in results.manufacturers"
            :key="manufacturer._id"
            :to="`${URL_MANUFACTURER_EDIT}/${manufacturer._id}`"
            @click="clearSearch"
          >
            {{ manufacturer.title }}
          </RouterLink>
        </div>

        <div v-if="results?.managers.length" :class="$style.resultsInner">
          <div>Managers:</div>
          <RouterLink
            v-for="manager in results.managers"
            :key="manager._id"
            :to="`${URL_MANAGER_EDIT}/${manager._id}`"
            @click="clearSearch"
          >
            {{ manager.firstName }} {{ manager.lastName }}
          </RouterLink>
        </div>
      </template>
    </div>
  </div>
</template>

<script setup lang="ts">
import { computed, ref, watch } from 'vue';

import { UiInput } from 'mhz-ui';
import { debounce } from 'perfect-debounce';

import IconSearch from '@/layout/icons/search.svg?component';
import { search } from '@/common/services';
import { URL_PRODUCT_EDIT } from '@/product/constants';
import { URL_CATEGORY_EDIT } from '@/category/constants';
import { URL_MANUFACTURER_EDIT } from '@/manufacturer/constants';
import { URL_MANAGER_EDIT } from '@/manager/constants';

const searchQuery = ref('');

const { data: results, refetch, isSuccess } = search(searchQuery, { enabled: false });

const debounced = debounce(async () => {
  refetch();
}, 300);

watch(
  () => searchQuery.value,
  async () => {
    if (searchQuery.value.length > 2) await debounced();
  }
);

function clearSearch() {
  searchQuery.value = '';
}

const isResults = computed(() => {
  return results.value
    ? results.value.categories.length +
        results.value.products.length +
        results.value.manufacturers.length +
        results.value.managers.length >
        0
    : false;
});
</script>

<style module lang="scss">
.container {
  position: relative;
  width: 320px;
}

.results {
  position: absolute;
  z-index: 2;
  display: flex;
  flex-direction: column;
  gap: 8px;
  width: 100%;
  max-height: 320px;
  padding: 16px;
  margin-top: 8px;
  overflow-y: auto;
  background-color: var(--color-white);
  border-radius: 16px;
  box-shadow: 0 0 16px -8px var(--color-black-transparent);
}

.resultsInner {
  display: flex;
  flex-direction: column;
  gap: 8px;
}
</style>
