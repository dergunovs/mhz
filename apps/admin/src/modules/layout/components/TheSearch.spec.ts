import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import TheSearch from './TheSearch.vue';

import { wrapperFactory, mockQueryReply } from '@/common/test';
import * as commonServices from '@/common/services';

const SEARCH_RESULTS = {
  products: [{ _id: '64d8d74249f373457450ccef', title: 'AMD Ryzen 9 7950X' }],
  categories: [],
};

const SMALL_QUERY = 'A';
const QUERY = 'AMD';

const refetch = vi.fn();

vi.spyOn(commonServices, 'search').mockImplementation(() => {
  return mockQueryReply<{ [key: string]: { _id: string }[] }>(SEARCH_RESULTS, refetch);
});

const search = dataTest('search');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(TheSearch, {});
});

enableAutoUnmount(afterEach);

describe('TheSearch', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(TheSearch)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('searches by watching input', async () => {
    expect(refetch).toBeCalledTimes(0);

    await wrapper.findComponent(search).setValue(SMALL_QUERY);

    expect(refetch).toBeCalledTimes(0);

    await wrapper.findComponent(search).setValue(QUERY);

    expect(refetch).toBeCalledTimes(1);
  });
});
