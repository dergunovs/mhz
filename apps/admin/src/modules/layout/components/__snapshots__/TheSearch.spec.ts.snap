// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TheSearch > matches snapshot 1`] = `"<div><input type="text" modelvalue="" searchscheme="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" results="[object Object]" issuccess="true" lang="en" data-test="search"></div>"`;
